### PR TITLE
change(web): enforce layout and validate only text attributes of subkeys 🐵

### PR DESCRIPTION
**Pull request by jahorton**
Previous versions of KeymanWeb would force all properties of subkeys to be a string type, in order to prevent type errors later in processing. However, layout properties are now numeric, and it was determined that they should not be customizable in any case.

This change enforces string type for text properties, and overwrites any layout properties with the defaults for subkeys.

Addresses https://github.com/keymanapp/keyman/pull/9664#discussion_r1354313379 and https://github.com/keymanapp/keyman/pull/9664#discussion_r1363445450.

@keymanapp-test-bot skip

### CODE REVIEW

File: web/src/engine/osk/src/input/gestures/browser/oskSubKey.ts [hunk 1]
```diff
@@ -24,12 +42,15 @@ export default class OSKSubKey extends OSKKey {
     let tKey = osk.getDefaultKeyObject();
```
Comment (jahorton) [inline web/src/engine/osk/src/input/gestures/browser/oskSubKey.ts:42]: References:

https://github.com/keymanapp/keyman/blob/2366653818820e11645716c597e386a3d322ca7b/web/src/engine/osk/src/visualKeyboard.ts#L519-L529

and

https://github.com/keymanapp/keyman/blob/2366653818820e11645716c597e386a3d322ca7b/common/web/keyboard-processor/src/keyboards/activeLayout.ts#L32-L38

**Review by mcdurdin — CHANGES_REQUESTED** [review]
This no longer validates types of a number of properties: `nextlayer`, `font`, 'fontsize', `sp`, `layer`, `default`.

While these are (mostly) not string types any longer, with this change we are less robust with bad input data than we were previously.

**Comment by jahorton** [comment]
Compared to the original... any entry among those that wasn't a string would be straight-up erased, with the exception of `sp`.  (To be precise, they'd be set to `undefined` due to the overwrite behavior.)

If you're looking for that behavior to be conceptually preserved... maybe a Map of key to expected `typeof` value would do the trick?

**Comment by mcdurdin** [comment]
> If you're looking for that behavior to be conceptually preserved... maybe a Map of key to expected `typeof` value would do the trick?

Something like that, yes. I think the purpose of the original was to cleanse our input types -- and that's valuable functionality to preserve. Of course I think this just scratches the surface of validation. 

Ideally it might emit console.warn or sentry logs on bogus keyboards.

Last but not least, might be clearer if this was an explicit keyboard-load-time function, which had the job of validating _all_ of the input keyboard props. Bigger job but if well designed could use that as a helpful compile toolchain function too, particularly for hand-crafted keyboards such as the CJK ones. (And not least, as a unit test for the compiler)

**Comment by jahorton** [comment]
The Developer build failure:  something went funky with the Sentry symbol upload.  Seems like the build itself was fine; t'was just an internet connectivity issue.

**Comment by mcdurdin** [comment]
> The Developer build failure: something went funky with the Sentry symbol upload. Seems like the build itself was fine; t'was just an internet connectivity issue.

When these intermittent failures occur, best to just rebuild.

File: common/web/keyboard-processor/src/keyboards/activeLayout.ts [hunk 2]
```diff
@@ -212,6 +233,51 @@ class ActiveKeyBase {
       rawKey.sp = Number.parseInt(rawKey.sp, 10) as ButtonClass;
     }
     rawKey.sp ||= 0; // The default button class.
+
+    // And now for generalized type validation. -----------------------------------------
+
+    // Object.entries does require Android 54... but we do polyfill within the Android app.  Should be 'fine'.
```
Comment (mcdurdin) [inline common/web/keyboard-processor/src/keyboards/activeLayout.ts:239]: destructuring is Chrome 49. So ... hope that is also polyfilled!

File: common/web/keyboard-processor/src/keyboards/activeLayout.ts [hunk 3]
```diff
@@ -23,6 +23,24 @@ function Enumerable(
     descriptor.enumerable = true;
 };
 
+const KeyTypesOfKeyMap = {
```
Comment (mcdurdin) [inline common/web/keyboard-processor/src/keyboards/activeLayout.ts:26]: Is there no way we can use the type data from common/web/types? This is just a bit WET

File: common/web/keyboard-processor/src/keyboards/activeLayout.ts [hunk 4]
```diff
@@ -212,6 +233,51 @@ class ActiveKeyBase {
       rawKey.sp = Number.parseInt(rawKey.sp, 10) as ButtonClass;
     }
     rawKey.sp ||= 0; // The default button class.
```
Comment (mcdurdin) [inline common/web/keyboard-processor/src/keyboards/activeLayout.ts:235]: We should be able to use the constant here rather than the number

**Review by mcdurdin — APPROVED** [review]

File: common/web/keyboard-processor/src/keyboards/activeLayout.ts (hunk 2 above)
Comment (jahorton) [reply common/web/keyboard-processor/src/keyboards/activeLayout.ts:239]: Destructuring is TS-down-compiled, so no worries there.  This isn't the first time it's been used in the code base.

From the actual TS compilation:

```javascript
        // And now for generalized type validation. -----------------------------------------
        // Object.entries does require Android 54... but we do polyfill within the Android app.  Should be 'fine'.
        for (var _i = 0, _a = Object.entries(KeyTypesOfKeyMap); _i < _a.length; _i++) {
            var _b = _a[_i], key = _b[0], value = _b[1];
            switch (value) {
```

No destructuring syntax is in the final product.

**Pull request merged**